Window Manger Refactor

natecraddock opened the pull request:
This is a large pull request so I don't expect each line of code to be closely scrutinized. The files that I care about most for review are:
* `layers.py` (new)
* `events.py` (new)
* `window.py`
* `components.py`

More details on the specifics of the changes are found in #28, but I'll give a brief overview here too:

The major changes here are:
1. The window manager has a stack of ComponentLayer objects. The window's `loop` function iterates through the layers drawing, passing events, and adding/removing layers as needed.
2. Event handling is moved to `events.py` from the Window class
3. The user input is no longer on a separate thread
4. The `layers.py` file contains ComponentLayer objects which are a usability layer between the window manager and the Component objects themselves. These layers allow for easily adding, updating, and changing settings on the components.

Closes #28

natecraddock committed:
UI: Refactor window input into a separate module

Create a zygrader/ui/input.py file to handle the input thread
separate from the window. Yay for code quality!

At the moment, resizing, the VIM header indicator, and the christmas
theme alternating events are broken. This will be fixed more with more
refactoring of the event loop.

natecraddock committed:
Window: Event loop beginnings

Add the basic structure of the new event loop. The new idea is to move event loops from component
create functions into the window. This makes it much simpler to add extra event types to be
handled by the window (Like adding tabs).

natecraddock committed:
Comment out resize and loop code

natecraddock committed:
Start work on creating layers

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Main menu registration

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Events: Remove separate input thread & cleanup

Rename input.py to events.py
Rename Input to EventManager (class)

At this point, zygrader is still in a broken state, but the beginnings
of the new window manager and event queue are working.

Because we now use halfdelay(1) for input, we no longer need
a separate input thread. This reduces a lot of code complexity.

natecraddock committed:
Window: Handle events in the main loop

Filtered lists now take input! Nothing happens yet, but we are getting
closer.

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Add events.py file

natecraddock committed:
Preferences: Connect preference observer to window and events

When preferences are changed, update the variables on the window and
event manager.

This is mostly untested, because the window manager doesn't allow entering
the preferences submenu anymore :)

natecraddock committed:
EventManager: move input preferences

Move input preferences from the window to the event manager.

natecraddock committed:
Cleanup: Rename window.events -> window.event_manager

natecraddock committed:
Add FunctionLayers

Allow closing windows (handled by wm!)

natecraddock committed:
Basics of coroutines

natecraddock committed:
Remove Tabs from the window manager

Tabs were a fun idea, but beyond the needed complexity of zygrader.
Zygrader needs to be kept within it's scope - a utility not a do-all.

If you need multiple zygrader windows open, there are utils for that
including tmux, terminal tabs, and even multiple terminal windows open.

natecraddock committed:
Add Popup and TextInput layers

natecraddock committed:
Cleanup: Imports

natecraddock committed:
Layers: Add a WorkerThread

This thread class will be used by the WaitPopup and Logger Layers.
It creates a thread for running long-duration code with a simple interface.

natecraddock committed:
MenuLayer layout tweaks

natecraddock committed:
Initial ListPopup layer

natecraddock committed:
Allow cancelling a wait popup

Add an option to wait popups to cancel the long-running operation.
Callers should check WaitPopup.was_canceled() before continuing.

natecraddock committed:
Cleanup: Remove unused code

natecraddock committed:
Add unregister_layer function

This is used to pop the top layer off the stack.
That said, should the functions be push_layer and pop_layer rather than
register_layer and unregister_layer?

natecraddock committed:
Layers: Add `canceled` to all layers; use in TextInput

The canceled property is useful enough that we should use it everywhere.

Also use in the login code so you can cancel at any point in the process.

natecraddock committed:
Cleanup: Remove old main menu code

natecraddock committed:
SearchableList base class

Share code wisely between the FilteredList and ListPopups.

Drawing & navigating the list works, but other actions crash.

natecraddock committed:
Add searching to SearchableList

natecraddock committed:
Finish new filtered list and list popup components

The component layers are now complete for the ScrollableList children.
The layers need to be written next.

natecraddock committed:
Remove old filtered list and list popup components

natecraddock committed:
Layers: Add RowHolder base class

natecraddock committed:
Begin preferences list popup redesign

natecraddock committed:
Rows: Add Toggle Row objects and use for preferences

Add an object that can be extended to toggle a row and call functions
when toggled.

natecraddock committed:
Rows: Add expandable rows

Currently only supports rows 2 deep

natecraddock committed:
Radio Buttons

natecraddock committed:
Cleanup: Simplify logic for Rows

Also add a generator for getting the index of the current row

natecraddock committed:
Toggles: Support an extra callback

natecraddock committed:
User: Replace old preferences menu with nested menu

Use the new layers module to make the code much more readable.
\o/ deleted code!

natecraddock committed:
Fix setting default text on text inputs

natecraddock committed:
Prep Lab Calc: Use new layers for popups

natecraddock committed:
Changelog: Use new layers system

natecraddock committed:
Admin Menu: Replace with new menu code

Only the first layer is replaced, so the submenus are still broken.
Fix the remove locks feature to work, and be more intuitive

natecraddock committed:
Window & Layers: Add rebuild and redraw flags

Control the redrawing and rebuilding of the event loop more closely
by flagging layers that need a rebuild. This means zygrader requires
even less processing power and won't erase/rebuild unless needed.

natecraddock committed:
Main Menu: Remove search prompt

No need to search the main menu.

natecraddock committed:
Handle back buttons properly

natecraddock committed:
Cleanup: Remove `blocking` from components

This will be in the Layers level

natecraddock committed:
Window: Only redraw windows that are visible

The redraw loop now only redraws when a layer is tagged for redraw.
If any layer is tagged, then all visible layers beneath it are also
tagged for a redraw, otherwise they will be cleared when the
screen is erased.

Commit:
Phythonify redraw check using `any`

natecraddock committed:
Cleanup: Class Manager menu

Use the new layout code for the class manager

natecraddock committed:
Admin: Add submission search menus & Path input layer

Add a new path input layer that doesn't allow confirmation until
the path is validated.

natecraddock committed:
Fix: Select back button on empty lists

natecraddock committed:
Merge branch 'wm-refactor' of github.com:natecraddock/zygrader into wm-refactor

natecraddock committed:
Use WaitPopup for grade puller

natecraddock committed:
Grade Puller: Use new layers instead of window for UI

This makes the section selector code much cleaner!

natecraddock committed:
Support variable arguments on row callbacks

natecraddock committed:
Refactor datetime spinners

natecraddock committed:
Grade Puller: Fix setting due dates

natecraddock committed:
Add new Logger layer

natecraddock committed:
Window: Remove old components code

Now that all components have moved to the layers.py file we can remove the
component management from the window entirely.

There are still some places that call these functions. These will be cleaned
up in future commits.

natecraddock committed:
Cleanup: Window instance

natecraddock committed:
Header: Update header drawing code

Attach headers to layers

natecraddock committed:
Fix: Wrong lab passed to submission search

natecraddock committed:
Grader: Begin moving grader to new layers

natecraddock committed:
Update View Student menu with new layer code

natecraddock committed:
Update versioning popup code

natecraddock committed:
Replace Popups, BoolPopups, and TextInputs with new layer system

natecraddock committed:
Window: Support layer resize events

natecraddock committed:
Fix: Missing default param on OptionsPopup

natecraddock committed:
Layers: Combine ListLayer and ListPopup classes

natecraddock committed:
Fix: Versioning downgrading user version

natecraddock committed:
Fix: Remove old header set code

natecraddock committed:
Layers: Cleanup commented code

natecraddock committed:
Grader: Show student list with auto update

natecraddock committed:
Email Manager: Use new layers for auto update

natecraddock committed:
Layers: Add Destroy functions, use for fs_watch

natecraddock committed:
Layers: Replace all remaining window.create_ calls with layers

natecraddock committed:
Fix: Broken string on flag note

natecraddock committed:
Cleanup: Component Constructors

natecraddock committed:
Add "Close" option to OptionsPopups

natecraddock committed:
Fix: Grader popup options

natecraddock committed:
Grader: Show submission details in popup with auto update

natecraddock committed:
Fix: Pair programming grading

The pair programming grader popup would open the regular student submission
popup due to an incorrect callback function.

natecraddock committed:
Fix: Pair programming students can be the same

Prevent selecting the same student twice for pair programming

teiknsmith commented on zygrader/class_manager.py:
```diff
@@ -1,4 +1,5 @@
 """Class Manager: Functions to manage zybooks classes"""
+from datetime import date
```
As far as I can tell this import is unused

teiknsmith replied:
```suggestion
```

teiknsmith commented on zygrader/grade_puller.py:
```diff
@@ -79,24 +84,30 @@ def read_canvas_csv(self):
         except PermissionError:
             msg = [
                 f"Could not open {path} for reading",
                 "Please have the owner of the file grand read permissions",
```
```suggestion
                "Please have the owner of the file grant read permissions",
```
Just found this small typo scanning through all the files.

teiknsmith commented on zygrader/user.py:
```diff
@@ -84,218 +92,137 @@ def login(window: ui.Window):
     if sign_in:
         login(window)
     else:
+        # TODO: Use the event manager to handle this case!
```
Seems like this should be resolved before we merge

teiknsmith commented on zygrader/user.py:
```diff
@@ -84,218 +92,137 @@ def login(window: ui.Window):
+    radio = PreferenceRadio("editor")
+    for editor_name in preferences.EDITORS:
+        radio.add_value(editor_name)
+    for editor_name in preferences.EDITORS:
+        row.add_row_radio(editor_name, radio)
```
To me it would feel more intuitive to have a `add_radio_group` function in `Row` that fills in the appropriate rows from the `Radio`, instead of needing to double for loop here

natecraddock replied:
I want to do this before we make 5.0.0 official, but after the merge.

teiknsmith commented on zygrader/user.py:
```diff
@@ -84,218 +92,137 @@ def login(window: ui.Window):
-        pref.select_fn()
-
+        radio.add_value(code)
+    for code in class_codes:
+        row.add_row_radio(code, radio)
```
Same as lines 204-205

teiknsmith commented on zygrader/ui/window.py:
```diff
@@ -56,208 +19,32 @@ def __init__(self, window, event: Event, component, custom_data):
 
     def __init__(self, callback, window_name):
-        Window.instance = self
+        Window.INSTANCE = self
         """Initialize screen and run callback function"""
```
Is there any reason the docstring is not at the top of this function?

natecraddock replied:
Just a simple mistake..

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -2,14 +2,14 @@
     def __init__(self):
-        # This determines if a component blocks layers beneath it completely
-        self.blocking = True
+        pass
```
If `__init__` doesn't do anything anymore, I would think just get rid of it

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -159,23 +156,22 @@ def resize(self, rows, cols):
+        super().__init__(height, width, title, message)
+        self.options = options
         self.index = len(options) - 1
         self.options_length = sum([len(o) for o in options]) + len(options) + 2
```
```suggestion
        self.set_options(options)
```

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -205,36 +206,40 @@ def last(self):
+
+    def set_include_date(self, include_date):
+        self.include_date = include_date
+
```
Including (or not) the date means the fields and format string need to update as well.
```suggestion
        self.__init_fields()
        self.__init_format_str()

```

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
-        self.dirty = True
+    @staticmethod
+    def create_line(index, text, color=0, sort_index=0):
+        return ScrollableList.Line(index, text, color, sort_index)
```
This method feel superfluous, creating a `ScrollableList.Line` directly doesn't seem that strange. I personally would replace calls to this function with a direct creation and remove this function.

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
-        except:
-            pass
+    def set_exit_text(self, text: str):
+        self._exit_text = text
```
This is doing nothing - I believe the intention is that it should modify `_lines[0]`

teiknsmith replied:
This was meant to just be on 678-679, I accidentally grabbed two lines prior as well, but they have nothing to do with this.
Well, I figured out that comments on individual lines get 3 lines of context before the line being commented on. Not a super fan of that.

natecraddock replied:
This is intentional, just not written correctly :)

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
+    def _filter(self):
+        # Use the filtered lines from the input search function.
+        # The search function should take a list and return a
+        # subset of the original list.
```
This comment feels like it's carried over from old code, it doesn't seem to match the current functionality of this method.

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
 
-        self.__fill_text(self.data)
-        self.window.noutrefresh()
+        # A special action (back, quit, etc.)
```
I'm pretty sure this should be yeeted, either that or I'm missing something entirely here.

teiknsmith replied:
Should just be line 698 referenced here, not sure why I picked up the other lines as well. I suppose it's okay to say they should be yeeted, but they already have been.

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
 
-        self.selected_index = 0
+        self._selected_index = 0
         self.set_scroll()
```
This (with line +741) is repeated multiple times and the functionality it drives at is a little unintuitive. We could abstract a `scroll_to_top` function for usage here and in the other search editing functions.

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
+        self.cols = cols
+
+        # List box
+        # TODO: Handle vertical sizing dependant on _searchable
```
This is a good TODO to resolve before the merge

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
+            self._selected_index = 1
+
+        # Draw the list lines
+        # TODO: Can we avoid slicing here?
```
Seems like a reasonable TODO, but not needed for this pull.

teiknsmith commented on zygrader/ui/components.py:
```diff
@@ -616,193 +621,237 @@ def _set_month_from_chars(self, newchar, recursed=False):
+
+        self.draw_title()
+
+        visible_lines = self._display_lines[self._scroll:self._scroll +
+                                            self.rows - ListPopup.V_PADDING]
+        for line_number, line in enumerate(visible_lines):
+            if line_number + self._scroll == self._selected_index:
+                add_str(self.window, Popup.PADDING + line_number, Popup.PADDING,
+                        f"{ScrollableList.SELECTED_PREFIX}{line.text}",
+                        curses.A_BOLD)
+            else:
+                add_str(self.window, Popup.PADDING + line_number, Popup.PADDING,
+                        f"{ScrollableList.UNSELECTED_PREFIX}{line.text}",
+                        curses.A_DIM)
```
This is very similar to `FilteredList`'s `draw`, and feels like at least some of the logic could be shared between the two in `ScrollableList`. It's different enough that I'm not sure on this, but think we should consider it.

natecraddock replied:
This is good, but I'd rather wait until after the refactor here too.

teiknsmith commented on zygrader/ui/events.py:
```diff
@@ -0,0 +1,233 @@
+
+UI_LEFT = 0
+UI_RIGHT = 1
+UI_CENTERED = 2
```
These aren't used anywhere. Do they have a future purpose, or should they be deleted?

teiknsmith commented on zygrader/ui/events.py:
```diff
@@ -0,0 +1,233 @@
+    def __init__(self, event_type, value, modifier=None):
+        self.type = event_type
+        self.value = value
+        self.modifier = modifier
```
`modifier` is not used anywhere anymore

teiknsmith replied:
So we should yeet it, and all references to it

teiknsmith commented on zygrader/ui/events.py:
```diff
@@ -0,0 +1,233 @@
+
+        # Input preferences
+        self.left_right_menu_nav = False
+        self.use_esc_back = False
```
Should these not be set using `update_preferences`? (and also `vim_mode` on line 65 if so)

teiknsmith commented on zygrader/ui/events.py:
```diff
@@ -0,0 +1,233 @@
+        # curses.doupdate()
+        return Event.HEADER_UPDATE
+
+    def disable_modes(self):
+        """Disable non-normal modes"""
+        self.insert_mode = False
+        self.mark_mode = False
```
This isn't used anywhere

natecraddock replied:
I had forgotten to call it when unregistering layers. Used now.

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+    """Execute a function in a separate thread and send data back to the owner ComponentLayer.
+
+    The data queue is for sending data from the worker thread function to the component.
+    The worker_fn should accept a single argument which is the queue.
```
These lines don't make much sense, they feel like they're related to old code.

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        self.redraw = True
+
+    def event_handler(self, event: Event, event_manager: EventManager):
+        pass
```
Should this `raise NotImplementedError` instead of just pass? Or are there cases where a Layer class doesn't need to implement `event_handler`?

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+
+    def set_text(self, text: str):
+        self.component.text = text
+        self.component.cursor_index = len(text)
```
I think this logic belongs with the Component in a `set_text` method (and then the component's `__init__` could make use of it too).

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        else:
+            prompt += [f"[Invalid {TYPE_STR}]"]
+        # super().set_prompt(prompt)
+        self.component.set_message(prompt)
```
The comment can be deleted, or used and the other line deleted.

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        row.do_action()
+
+
+class ListLayer(ComponentLayer, Row, PopupLayer):
```
It feels very strange that a ListLayer *is a* Row - composition feels more intuitive here

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+
+        self.__subrows: List[self.__class__] = []
+        self.__callback_fn = None
+        self._callback_args = None
```
Is there a reason for under instead of dunder for `callback_args`?

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+            if self._callback_args:
+                self.__callback_fn(*self._callback_args)
+            else:
+                self.__callback_fn()
```
If callback args is defaulted to `()`, this should be able to become
```suggestion
            self.__callback_fn(*self._callback_args)
```

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        self.build_string_lines(text_rows, self)
+        self.component.set_lines(text_rows)
+
+    # TODO: move somewhere else?
```
This seems well enough here, but could be moved to a util group if wanted

teiknsmith replied:
Again, this picked up some lines ahead of what I want to comment on - the TODO is tied to lines after this.

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+            self.component.delchar()
+            self.rebuild = True
+        elif event.type == Event.ESC and event_manager.use_esc_back:
+            # TODO: Handle this event by the window manager?
```
This section seems good as-is
```suggestion
```

teiknsmith replied:
This TODO also applied to lines not shown in the preview

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        return self.component.get_time()
+
+
+class Radio:
```
`RadioGroup` feels like a more intuitive name for this - as is, it looks like it is similar to `Toggle`, which is just one row, but this class manages a group of toggleable radio items.

teiknsmith commented on zygrader/user.py:
```diff
@@ -84,218 +92,137 @@ def login(window: ui.Window):
-    """Set the user's default editor to the selected editor"""
-    editor = list(preferences.EDITORS.keys())[editor_index]
-    preferences.set(pref_name, editor)
+class PreferenceRadio(ui.layers.Radio):
```
If `Radio` becomes `RadioGroup`, `PreferenceRadio` should also become `PreferenceRadioGroup`
(`RadioGroup` commented on in `layers.py` line 367)

teiknsmith commented on zygrader/ui/layers.py:
```diff
@@ -0,0 +1,634 @@
+        self.__set_fn(self.__name, value)
+
+
+class Toggle:
```
`Toggle` is using a very strange combination of function passing and inheritance/overriding functions.
It seems the same functionality could be accomplished with this `Toggle` class being a simple interface with no data members and functions to override:
- is_toggled(): Bool
- toggle(): void

teiknsmith reviewed (changes requested):
A few comments that don't fit well with any single group of lines:
- Updating the header to cycle the colors needs some fixes. It's obvious that this was considered and noted, but should be finished. Some particular places to fix/resolve:
    - `window.py` lines 215-217
    - `events.py` lines 115-118
    - `events.py` lines 181-182
- Unused imports. I commented on one specifically, but found multiple while running through, and some aren't available for comment on this commit. We should go through all the files and take care of these.
- `zygrader/utils.py` line 211-212 still uses the old window system
- `versioning.py` has a few function names that conflict with what they do (and even their docstrings). It also feels strange that there are two entry points to `do_versioning` (which is a pretty vague name) and then to separately `show_versioning_message(s)` - I see why this was done, but it would be neat if we could find a way to simplify that.
- The readme needs a big look-at in the `Code Documentation` section.

natecraddock committed:
Cleanup: Reorder and remove imports

Reorder imports alphabetically, and remove unused imports.

natecraddock committed:
Fix: Zybooks TOC fetch using old code

natecraddock committed:
Cleanup: Versioning code naming and comments

natecraddock commented:
Thanks for the thorough review!
* All imports are now cleaned & organized
* I think the readme documentation can be done as a step after the merge
* There may be a way to combine the versioning code into a single function again, but I think that would be best left until after the refactor. I did update the name of the function though.

natecraddock committed:
Cleanup: Comments, unused code, organization

General cleanups to the ui code to make things readable. No functional
changes here.

natecraddock committed:
Add zygrader quit event. Use for logout

Add an event that can be used to exit the application from
any point in the software.

natecraddock committed:
Cleanup: Remove unused function

natecraddock committed:
Cleanup: remove TODO comment

natecraddock committed:
Cleanup: TextInput set_text method

natecraddock committed:
Cleanup: Remove TODO comment

natecraddock committed:
ListPopups: Add search prompt

Add the ability to search and sort the text in a list popup

natecraddock commented:
Another thing, I don't care about keeping the the cycling header colors. I know @SeaSaltTree  and @mistBourne1300 have been working with headers & colors. Do either of you have an opinion on removing the header color alternating on keypresses?

natecraddock committed:
Disable VIM modes when leaving layers

Always exit insert and visual mode when exiting or creating a layer.

natecraddock committed:
Fix: Password saved after saying "No" to save password prompt

The preference was never set to False if a user already had True
in their preferences.

natecraddock committed:
Text Input Areas: Show cursor

natecraddock committed:
Cleanup: Move list scroll to top code into separate function

natecraddock committed:
Merge branch 'master' into wm-refactor

natecraddock committed:
Raise version number to 5.0.0

natecraddock committed:
Readme: Update code docs

The documentation is very outdated after the refactor. It was decided to remove
the docs and just keep cleaner comments in the code itself.

natecraddock committed:
Cleanup: Row callback args

natecraddock committed:
Cleanup: Header drawing code

Remove a few TODOs and optimize when the header is redrawn

natecraddock committed:
Optimize redraws for layers

Only redraw if an event was not Event.NONE

natecraddock committed:
Cleanup: Make Toggle an abstract class

Toggle was a strange mix of callbacks and inheritance. Let's just make it
an abstract class that defines an interface for how toggles should behave.

natecraddock committed:
ListLayer: Remove "is a" relationship with Rows

A ListLayer now contains a reference to a row. This increases some code duplication,
but it makes things easier to read.

natecraddock committed:
Fix: Colors on list popups

natecraddock commented:
@spencerteiknsmith I've addressed every concern you raised. I also
* Merged with master
* Removed the docs
* Updated the version string to `5.0.0`

Once you are happy with everything just go ahead and accept the pull request. I'd like to do the merge manually, so don't do it through github.

SeaSaltTree commented:
I like having colors, but they don't have to alternate.

SeaSaltTree commented:
Yesterday I was able to grade and lock emails just fine on the refactor branch. Today, normal zygrader (version 4.9) works. In the virtual environment, under the master branch, it works. But in the virtual environment, under the wm-refactor branch, after checking it out and doing git pull, whenever I click email and grade>lab 6 it goes wack. I'm not sure whether this is just me.
<img width="500" alt="wm-refactor" src="https://user-images.githubusercontent.com/70764441/97515286-5e01f300-1956-11eb-91a5-e1ad74fbf27e.PNG">

natecraddock commented:
@SeaSaltTree thanks, just fixed that bug.
